File: server/app/monitor.py
"""Real-Time Monitoring.

Tracks the lifecycle of a generated signal against live prices:
WAITING -> ENTRY_TRIGGERED -> IN_PROFIT -> TAKE_PROFIT / STOP_LOSS / INVALIDATED.
In-memory only; swap for a database-backed store for persistence across restarts.
"""
from __future__ import annotations

import itertools
import time
from dataclasses import dataclass, field

_id_counter = itertools.count(1)


@dataclass
class MonitoredTrade:
    id: int
    symbol: str
    direction: str
    entry: float
    stop_loss: float
    take_profit_1: float
    take_profit_2: float
    status: str = "WAITING"
    created_at: float = field(default_factory=time.time)
    updated_at: float = field(default_factory=time.time)
    history: list[dict] = field(default_factory=list)

    def _transition(self, new_status: str, price: float):
        self.status = new_status
        self.updated_at = time.time()
        self.history.append({"status": new_status, "price": price, "at": self.updated_at})

    def update_price(self, price: float):
        long = self.direction == "long"
        if self.status == "WAITING":
            triggered = price >= self.entry if long else price <= self.entry
            if triggered:
                self._transition("ENTRY_TRIGGERED", price)
            return

        if self.status in ("ENTRY_TRIGGERED", "IN_PROFIT"):
            hit_tp = price >= self.take_profit_1 if long else price <= self.take_profit_1
            hit_sl = price <= self.stop_loss if long else price >= self.stop_loss
            if hit_sl:
                self._transition("STOP_LOSS", price)
            elif hit_tp:
                self._transition("TAKE_PROFIT", price)
            else:
                in_profit = price > self.entry if long else price < self.entry
                if in_profit and self.status != "IN_PROFIT":
                    self._transition("IN_PROFIT", price)


class TradeMonitor:
    def __init__(self):
        self._trades: dict[int, MonitoredTrade] = {}

    def track(self, symbol: str, direction: str, entry: float, stop_loss: float, tp1: float, tp2: float) -> MonitoredTrade:
        trade = MonitoredTrade(id=next(_id_counter), symbol=symbol, direction=direction, entry=entry, stop_loss=stop_loss, take_profit_1=tp1, take_profit_2=tp2)
        self._trades[trade.id] = trade
        return trade

    def update_price(self, symbol: str, price: float):
        for trade in self._trades.values():
            if trade.symbol == symbol and trade.status not in ("TAKE_PROFIT", "STOP_LOSS", "INVALIDATED"):
                trade.update_price(price)

    def active(self) -> list[MonitoredTrade]:
        return [t for t in self._trades.values() if t.status not in ("TAKE_PROFIT", "STOP_LOSS", "INVALIDATED")]

    def all(self) -> list[MonitoredTrade]:
        return list(self._trades.values())


trade_monitor = TradeMonitor()

File: server/src/monitor.ts
/**
 * Real-Time Monitoring.
 *
 * Tracks the lifecycle of a generated signal against live prices:
 * WAITING -> ENTRY_TRIGGERED -> IN_PROFIT -> TAKE_PROFIT / STOP_LOSS / INVALIDATED.
 * In-memory only; swap for a database-backed store for persistence across restarts.
 */

let idCounter = 1;

export type TradeStatus = "WAITING" | "ENTRY_TRIGGERED" | "IN_PROFIT" | "TAKE_PROFIT" | "STOP_LOSS" | "INVALIDATED";

export interface HistoryEntry {
  status: TradeStatus;
  price: number;
  at: number;
}

export class MonitoredTrade {
  id: number;
  symbol: string;
  direction: "long" | "short";
  entry: number;
  stop_loss: number;
  take_profit_1: number;
  take_profit_2: number;
  status: TradeStatus = "WAITING";
  created_at: number = Date.now() / 1000;
  updated_at: number = Date.now() / 1000;
  history: HistoryEntry[] = [];

  constructor(
    id: number,
    symbol: string,
    direction: "long" | "short",
    entry: number,
    stopLoss: number,
    tp1: number,
    tp2: number
  ) {
    this.id = id;
    this.symbol = symbol;
    this.direction = direction;
    this.entry = entry;
    this.stop_loss = stopLoss;
    this.take_profit_1 = tp1;
    this.take_profit_2 = tp2;
  }

  private transition(newStatus: TradeStatus, price: number) {
    this.status = newStatus;
    this.updated_at = Date.now() / 1000;
    this.history.push({ status: newStatus, price, at: this.updated_at });
  }

  updatePrice(price: number) {
    const long = this.direction === "long";
    if (this.status === "WAITING") {
      const triggered = long ? price >= this.entry : price <= this.entry;
      if (triggered) this.transition("ENTRY_TRIGGERED", price);
      return;
    }

    if (this.status === "ENTRY_TRIGGERED" || this.status === "IN_PROFIT") {
      const hitTp = long ? price >= this.take_profit_1 : price <= this.take_profit_1;
      const hitSl = long ? price <= this.stop_loss : price >= this.stop_loss;
      if (hitSl) {
        this.transition("STOP_LOSS", price);
      } else if (hitTp) {
        this.transition("TAKE_PROFIT", price);
      } else {
        const inProfit = long ? price > this.entry : price < this.entry;
        if (inProfit && this.status !== "IN_PROFIT") this.transition("IN_PROFIT", price);
      }
    }
  }
}

export class TradeMonitor {
  private trades = new Map<number, MonitoredTrade>();

  track(
    symbol: string,
    direction: "long" | "short",
    entry: number,
    stopLoss: number,
    tp1: number,
    tp2: number
  ): MonitoredTrade {
    const trade = new MonitoredTrade(idCounter++, symbol, direction, entry, stopLoss, tp1, tp2);
    this.trades.set(trade.id, trade);
    return trade;
  }

  updatePrice(symbol: string, price: number) {
    for (const trade of this.trades.values()) {
      if (trade.symbol === symbol && !["TAKE_PROFIT", "STOP_LOSS", "INVALIDATED"].includes(trade.status)) {
        trade.updatePrice(price);
      }
    }
  }

  active(): MonitoredTrade[] {
    return [...this.trades.values()].filter(
      (t) => !["TAKE_PROFIT", "STOP_LOSS", "INVALIDATED"].includes(t.status)
    );
  }

  all(): MonitoredTrade[] {
    return [...this.trades.values()];
  }
}

export const tradeMonitor = new TradeMonitor();
